com.unity.shadergraph/Editor/Generation/Processors/Generator.cs

```diff
@@ -97,56 +100,77 @@ void BuildShader()
-            m_GraphData.CollectShaderProperties(shaderProperties, m_Mode);
+            m_GraphData.CollectShaderProperties(allShaderProperties, m_Mode);
             m_GraphData.CollectShaderKeywords(shaderKeywords, m_Mode);
 
             if (m_GraphData.GetKeywordPermutationCount() > ShaderGraphPreferences.variantLimit)
             {
                 m_GraphData.AddValidationError(m_OutputNode.objectId, ShaderKeyword.kVariantLimitWarning, Rendering.ShaderCompilerMessageSeverity.Error);
 
-                m_ConfiguredTextures = shaderProperties.GetConfiguredTexutres();
+                m_ConfiguredTextures = allShaderProperties.GetConfiguredTexutres();
                 m_Builder.AppendLines(ShaderGraphImporter.k_ErrorShader);
             }
 
+            // Collect properties from the nodes
             foreach (var activeNode in activeNodeList.OfType<AbstractMaterialNode>())
-                activeNode.CollectShaderProperties(shaderProperties, m_Mode);
+                activeNode.CollectShaderProperties(allShaderProperties, m_Mode);
 
-            // Collect excess shader properties from the TargetImplementation
+            // Collect properties from the TargetImplementation
             foreach (var target in m_Targets)
             {
                 // TODO: Setup is required to ensure all Targets are initialized
                 // TODO: Find a way to only require this once
                 TargetSetupContext context = new TargetSetupContext();
                 target.Setup(ref context);
 
-                target.CollectShaderProperties(shaderProperties, m_Mode);
+                // flag all properties as being required by this target
+                allShaderProperties.BeginTargetCollection(target);
+                target.CollectShaderProperties(allShaderProperties, m_Mode);
+                allShaderProperties.EndTargetCollection(target);
             }
 
+            // set the property collector to read only
+            // (to ensure no rogue target or pass starts adding more properties later..)
+            allShaderProperties.SetReadOnly();
+
             m_Builder.AppendLine(@"Shader ""{0}""", m_Name);
             using (m_Builder.BlockScope())
             {
-                GenerationUtils.GeneratePropertiesBlock(m_Builder, shaderProperties, shaderKeywords, m_Mode);
+                GenerationUtils.GeneratePropertiesBlock(m_Builder, allShaderProperties, shaderKeywords, m_Mode);
 
                 for (int i = 0; i < m_Targets.Length; i++)
                 {
                     TargetSetupContext context = new TargetSetupContext(m_assetCollection);
+                    var target = m_Targets[i];
 
                     // Instead of setup target, we can also just do get context
-                    m_Targets[i].Setup(ref context);
+                    target.Setup(ref context);
+
+                    var subShaderProperties = GetSubShaderPropertiesForTarget(target, m_GraphData, m_Mode, m_OutputNode);
```

Comment:
Now when building subshaders for a Target, we generate the set of properties pertaining to that subshader, and use it to build the shared HLSL declaration of those properties.